### PR TITLE
SALTO-3080 Add template static file support to resolver

### PR DESCRIPTION
**Pull request by adfineberg**
1. move RestoreValuesFunc, restoreChangeElement and restoreValues to adapter-component
2. add template static file logic to restore
3. fix resolve logic for template static file

---

_Additional context for reviewer_

---
_Release Notes_: 
_Replace me with a short sentence that describes the effect of this change on Salto users_

---
_User Notifications_: 
_Replace me with a short sentence that describes changes that will appear in NaCls and are not caused by user actions (e.g. a new annotation, field values that are converted to references, etc). Hidden changes should not be listed._

### CODE REVIEW

File: packages/adapter-components/src/restore_utils.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      if (isEqualResolvedValues(refValue, value)) {
+        return ref
+      }
+    }
+
+    const file = allStaticFilesPaths.get(path.getFullName())
+    if (file !== undefined) {
+      if (file.isTemplate) {
+        return parserUtils.templateExpressionToStaticFile(value, file.filepath)
+      }
```
Comment (edenhassid) [inline packages/adapter-components/src/restore_utils.ts:106]: everything is moved from the adapter-utils/utils except for this part which is added

File: packages/adapter-components/test/restore_utils.test.ts
Comment (edenhassid) [inline packages/adapter-components/test/restore_utils.test.ts:1]: copied from adapter-utils

File: packages/adapter-utils/src/utils.ts [hunk 2]
```diff
@@ -539,121 +534,6 @@ export type ResolveValuesFunc = <T extends Element>(
   allowEmpty?: boolean,
 ) => Promise<T>
 
-export type RestoreValuesFunc = <T extends Element>(
```
Comment (ori-moisis) [inline packages/adapter-utils/src/utils.ts:542]: why was this moved?

File: packages/adapter-components/src/resolve_utils.ts [hunk 3]
```diff
@@ -32,21 +32,33 @@ import { parserUtils } from '@salto-io/parser'
...
     if (isStaticFile(value)) {
       if (value.isTemplate) {
-        return parserUtils.staticFileToTemplateExpression(value)
+        const templateExpression = await parserUtils.staticFileToTemplateExpression(value)
+        if (templateExpression)
+          templateExpression.parts = await Promise.all(
+            templateExpression?.parts.map(async part =>
+              isReferenceExpression(part) ? resolveReferenceExpression(part) : part,
+            ),
+          )
```
Comment (ori-moisis) [inline packages/adapter-components/src/resolve_utils.ts:60]: if we are starting to resolve template expressions here, we should be consistent about it.
IIUC this code will only resolve the references in template expressions that come from static files, but it won't resolve references in template expressions that came from NaCl

File: packages/adapter-components/src/restore_utils.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      const refValue = await getLookUpName({ ref, field, path, element: targetElement })
+      if (isEqualResolvedValues(refValue, value)) {
+        return ref
+      }
+    }
+
+    const file = allStaticFilesPaths.get(path.getFullName())
+    if (file !== undefined) {
+      if (file.isTemplate) {
+        return parserUtils.templateExpressionToStaticFile(value, file.filepath)
```
Comment (ori-moisis) [inline packages/adapter-components/src/restore_utils.ts:105]: I might be missing something, but it seems like there is no check that the current value is actually supposed to be a template static file (i.e - what if it is just a string and not a template expression?)
won't that be an issue? if the `value` is just a string, won't it cause the code to crash?

File: packages/adapter-components/src/restore_utils.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      }
+      const content = file.encoding === 'binary' ? value : Buffer.from(value, file.encoding)
```
Comment (ori-moisis) [inline packages/adapter-components/src/restore_utils.ts:107]: similar issue (though probably less prone to crashing) - should we check the value is a buffer / string before we make it a buffer and put it in a static file?

File: packages/adapter-components/src/resolve_utils.ts (hunk 3 above)
Comment (edenhassid) [reply packages/adapter-components/src/resolve_utils.ts:60]: references from template expressions are also resolved here, the `valuesReplacer` is called from `transformElement` which if I understand correctly calls `valuesReplacer` for all the values in the template expression at some point.

File: packages/adapter-components/src/restore_utils.ts (hunk 4 above)
Comment (edenhassid) [reply packages/adapter-components/src/restore_utils.ts:105]: I'll add a test that value is indeed a template expression

File: packages/adapter-components/src/resolve_utils.ts (hunk 3 above)
Comment (ori-moisis) [reply packages/adapter-components/src/resolve_utils.ts:60]: I also thought so, but if that is true, then there is no need to have this code here

Comment (edenhassid) [reply packages/adapter-components/src/resolve_utils.ts:60]: why if it is true there is no need to have this code here? we resolve a template static file to template expression and then in this case we don't go inside the template expression like we do if it just appears in the nacl

Comment (edenhassid) [reply packages/adapter-components/src/resolve_utils.ts:62]: discussed IRL

File: packages/adapter-components/src/restore_utils.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        return ref
+      }
+    }
+
+    const file = allStaticFilesPaths.get(path.getFullName())
+    if (file !== undefined) {
+      if (file.isTemplate && isTemplateExpression(value)) {
+        return parserUtils.templateExpressionToStaticFile(value, file.filepath)
+      }
+      if (isBuffer(value) || isString(value)) {
```
Comment (ori-moisis) [inline packages/adapter-components/src/restore_utils.ts:108]: what if `file.isTemplate` and `value` is a string?
we would not enter the first condition, and we would enter this one, so IIUC we would end up creating a template static file (because we take the encoding from the original file) that isn't actually a template

Comment (edenhassid) [reply packages/adapter-components/src/restore_utils.ts:108]: I think its valid
